src/utils/units.cpp

```diff
@@ -0,0 +1,115 @@
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#include "units.hpp"
+
+namespace ArtemisUtils {
+
+constexpr Real Msolar = 1.988416e33;
+constexpr Real AU = 1.495978707e13;
+constexpr Real Year = 31536000;
+constexpr Real parsec = 3.0857e18;
+constexpr Real Rjup = 6.991100e6;
+constexpr Real Mjup = 1.8982e30;
+
+Units::Units(ParameterInput *pin, std::shared_ptr<StateDescriptor> pkg) {
+  std::string unit_system_str = pin->GetOrAddString("artemis/units", "type", "scalefree");
+  if (unit_system_str == "scalefree") {
+    unit_system_ = UnitSystem::scalefree;
+  } else if (unit_system_str == "cgs") {
+    unit_system_ = UnitSystem::cgs;
+  } else {
+    PARTHENON_FAIL("Unit system not recognized! Valid choices are [scalefree, cgs]");
+  }
+
+  if (unit_system_ == UnitSystem::scalefree) {
+    length_ = 1.;
+    time_ = 1.;
+    mass_ = 1.;
+  } else {
+    std::string unit_specifier = pin->GetOrAddString("artemis", "specifier", "base");
+    if (unit_specifier == "base") {
+      length_ = pin->GetReal("artemis", "length");
+      time_ = pin->GetReal("artemis", "time");
+      mass_ = pin->GetReal("artemis", "mass");
+    } else if (unit_specifier == "ppd") {
+      length_ = AU;
+      mass_ = Msolar;
+      time_ = Year / (2. * M_PI);
+    } else {
+      PARTHENON_FAIL("Unit specifier not recognized!");
+    }
+  }
+
+  // Remaining conversion factors
+  energy_ = std::pow(length_, 2) * mass_ * std::pow(time_, -2);
+  number_density_ = std::pow(length_, -3);
+
+  // Store everything necessary in params for usage in analysis
+  pkg->AddParam("unit_system", unit_system_);
+  pkg->AddParam("length", length_);
+  pkg->AddParam("time", time_);
+  pkg->AddParam("mass", mass_);
+}
+
+Constants::Constants(Units &units) {
+  if (units.GetUnitSystem() == UnitSystem::scalefree) {
+    G_ = 1.;
+    kb_ = 1.;
+    c_ = 1.;
+    h_ = 1.;
+    amu_ = 1.;
+    eV_ = 1.;
+    Msolar_ = 1.;
+    AU_ = 1.;
+    pc_ = 1.;
+    Year_ = 1.;
+  } else if (units.GetUnitSystem() == UnitSystem::cgs) {
+    parthenon::constants::PhysicalConstants<parthenon::constants::CGS> pc;
+    G_ = pc.gravitational_constant;
+    kb_ = pc.kb;
+    c_ = pc.c;
+    h_ = pc.h;
+    amu_ = pc.amu;
+    eV_ = pc.eV;
+    Msolar_ = Msolar;
+    AU_ = AU;
+    Rjup_ = Rjup;
+    Mjup_ = Mjup;
+    pc_ = parsec;
+    Year_ = Year;
```

Comment:
Sorry, one more constant that is being used is the radiation constant and optionally the steffan-boltzman constant.

Reply:
Good point, `ar` added